src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java

```diff
@@ -93,4 +99,139 @@ public static String toQueryString(Map<String, String> params) {
+
+        // make sure each leg matches
+        for (int i = 0; i < previousItinerary.legs.size(); i++) {
+            Leg previousItineraryLeg = previousItinerary.legs.get(i);
+            Leg newItineraryLeg = newItinerary.legs.get(i);
+
+            if (!legsMatch(previousItineraryLeg, newItineraryLeg)) return false;
+        }
+
+        // if this point is reached, the itineraries are assumed to match
+        return true;
+    }
+
+    /**
+     * Check whether a new leg of an itinerary matches the previous itinerary leg for the purposes of trip monitoring.
+     */
+    private static boolean legsMatch(Leg previousItineraryLeg, Leg newItineraryLeg) {
+        // for now don't analyze non-transit legs
+        if (!previousItineraryLeg.transitLeg) return true;
+
+        // make sure the same from/to stop are being used
+        if (
+            !stopsMatch(previousItineraryLeg.from, newItineraryLeg.from) ||
+                !stopsMatch(previousItineraryLeg.to, newItineraryLeg.to)
+        ) {
+            return false;
+        }
+
+        // make sure the transit service is the same as perceived by the customer
+        if (
+            !equalsOrPreviousWasNull(previousItineraryLeg.mode, newItineraryLeg.mode) ||
+                !equalsIgnoreCaseOrPreviousWasEmpty(previousItineraryLeg.agencyName, newItineraryLeg.agencyName) ||
+                !equalsIgnoreCaseOrPreviousWasEmpty(
+                    previousItineraryLeg.routeLongName,
+                    newItineraryLeg.routeLongName
+                ) ||
+                !equalsIgnoreCaseOrPreviousWasEmpty(
+                    previousItineraryLeg.routeShortName,
+                    newItineraryLeg.routeShortName
+                ) ||
+                !equalsIgnoreCaseOrPreviousWasEmpty(
+                    previousItineraryLeg.headsign,
+                    newItineraryLeg.headsign
+                ) ||
+                !equalsOrPreviousWasNull(
+                    previousItineraryLeg.interlineWithPreviousLeg,
+                    newItineraryLeg.interlineWithPreviousLeg
+                )
+        ) {
+            return false;
+        }
+
+        // make sure the transit trips are scheduled for the same time of the day
+        if (
+            !timeOfDayMatches(
+                previousItineraryLeg.getScheduledStartTime(),
+                newItineraryLeg.getScheduledStartTime()
+            ) || !timeOfDayMatches(
+                previousItineraryLeg.getScheduledEndTime(),
+                newItineraryLeg.getScheduledEndTime()
+            )
+        ) {
+            return false;
+        }
+
+        // if this point is reached, the legs are assumed to match
+        return true;
+    }
+
+    /**
+     * Checks whether two stops (OTP Places) match for the purposes of matching itineraries
+     */
+    private static boolean stopsMatch(Place stopA, Place stopB) {
+        // stop names must match
+        if (!stopA.name.equalsIgnoreCase(stopB.name)) return false;
```

Comment:
I expect you've already thought of this, but it might be good to provide a comment here explaining that OTP requires that stop name is not null (I don't actually know if that's true).

Reply:
Fixed in ffc3b49b3ad576d936b04758d572efdf3f217ab2